designs/2021-suppression-support/README.md

```diff
@@ -0,0 +1,302 @@
+
+For external suppression (or configuration files in ESLint), there are no good places to put a justification. We may put something general as a justification, like "Globally disabled".
+
+For example:
+
+```json
+{"kind": "external", "justification": "Globally disabled"}
+```
+
+If a line is suppressed by multiple rules, a suppression list would be preferred to gather all suppression information:
+
+```json
+suppressions: [
+  {"kind": "inSource", "justification": "Justification message 1"},
+  {"kind": "inSource", "justification": "Justification message 2"},
+  {"kind": "external", "justification": "Globally disabled"}
+]
+```
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+### Goal
+
+1.	Keep the current behavior of ESLint, i.e., errors/warnings will not be shown on the command line if the related rules are disabled, unless `--no-inline-config` is used.
+2.	When the flag `--output-suppression` is used, the violations (problems) that disabled by inline comments will not be removed.
+3.	Add suppression info in the output of ESLint
+    1. For inSource suppressions, we generate suppression info where `kind` is `inSource` and `justification` is the description in the directive comments (words after two or more consecutive dashes mentioned above).
+    
+        For example:
+        
+        ```javascript
+        /* eslint-disable no-undef -- Justification example */
+        a = 1;
+        /* eslint-enable no-undef */
+        ```
+        ```javascript
+        console.log("foo"); // eslint-disable-line no-console -- Justification example
+        ```
+        ```javascript
+        // eslint-disable-next-line no-console -- Justification example
+        console.log("bar");
+        ```
+
+        Output:
+
+        ```json
+        {"kind": "inSource", "justification": "Justification example"}
+        ```
+    
+    2. For external suppressions, we generate suppression info where `kind` is `external` and `justification` is a certain sentence such as "Globally disabled".
+
+        For example:
+
+        ```json
+        rules: {
+          "no-undef": "off"
+        }
+        ```
+
+        Output:
+
+        ```json
+        {"kind": "external", "justification": "Globally disabled"}
+        ```
+
+### General Design
+
+We are proposing to add a new CLI option `--output-suppression` to output suppressions without prompting disabled errors/warnings.
+
+![Design diagram](design_diagram.png)
+
+Input:
+
+```
+.\node_modules\.bin\eslint.cmd 1.js --output-suppression -f @microsoft/eslint-formatter-sarif -o 1.sarif
+```
+
+Output:
+
+![Suppression example](suppression_example.png)
```

Comment:
It would be helpful to show the output from a ESLint with the JSON formatter. The readers of this document aren’t going to be familiar with SARIF. And please include that code online rather than as an image.

Reply:
Updated. Thanks for the suggestion!